Step 1: Load Dataset

In [4]:
import pandas as pd

#csv path below, must be edited to match where dataset is in user library
df = pd.read_csv('/Users/brandonacosta/Documents/GitHub/text-analytics-spring-2026/assignment_1/data/raw/BA_AirlineReviews.csv')

In [5]:
%matplotlib inline
import matplotlib.pyplot as plt

In [6]:
df.head()

Unnamed: 0  OverallRating                              ReviewHeader  \
0           0            1.0    "Service level far worse then Ryanair"   
1           1            3.0  "do not upgrade members based on status"   
2           2            8.0             "Flight was smooth and quick"   
3           3            1.0             "Absolutely hopeless airline"   
4           4            1.0        "Customer Service is non existent"   

               Name            Datetime  VerifiedReview  \
0           L Keele  19th November 2023            True   
1      Austin Jones  19th November 2023            True   
2        M A Collie  16th November 2023           False   
3        Nigel Dean  16th November 2023            True   
4  Gaylynne Simpson  14th November 2023           False   

                                          ReviewBody TypeOfTraveller  \
0  4 Hours before takeoff we received a Mail stat...  Couple Leisure   
1  I recently had a delay on British Airways from...        Business   
2  Boarded on time, but it took ages to get to th...  Couple Leisure   
3  5 days before the flight, we were advised by B...  Couple Leisure   
4  We traveled to Lisbon for our dream vacation, ...  Couple Leisure   

         SeatType                      Route      DateFlown  SeatComfort  \
0   Economy Class        London to Stuttgart  November 2023          1.0   
1   Economy Class         Brussels to London  November 2023          2.0   
2  Business Class  London Heathrow to Dublin  November 2023          3.0   
3   Economy Class           London to Dublin  December 2022          3.0   
4   Economy Class           London to Lisbon  November 2023          1.0   

   CabinStaffService  GroundService  ValueForMoney Recommended Aircraft  \
0                1.0            1.0            1.0          no      NaN   
1                3.0            1.0            2.0          no     A320   
2                3.0            4.0            3.0         yes     A320   
3                3.0            1.0            1.0          no      NaN   
4                1.0            1.0            1.0          no      NaN   

   Food&Beverages  InflightEntertainment  Wifi&Connectivity  
0             NaN                    NaN                NaN  
1             1.0                    2.0                2.0  
2             4.0                    NaN                NaN  
3             NaN                    NaN                NaN  
4             1.0                    1.0                1.0

Step 2: Explore the Dataset

In [8]:
df.shape

(3701, 20)

In [9]:
df.columns

Index(['Unnamed: 0', 'OverallRating', 'ReviewHeader', 'Name', 'Datetime',
       'VerifiedReview', 'ReviewBody', 'TypeOfTraveller', 'SeatType', 'Route',
       'DateFlown', 'SeatComfort', 'CabinStaffService', 'GroundService',
       'ValueForMoney', 'Recommended', 'Aircraft', 'Food&Beverages',
       'InflightEntertainment', 'Wifi&Connectivity'],
      dtype='object')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3701 entries, 0 to 3700
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             3701 non-null   int64  
 1   OverallRating          3696 non-null   float64
 2   ReviewHeader           3701 non-null   object 
 3   Name                   3701 non-null   object 
 4   Datetime               3701 non-null   object 
 5   VerifiedReview         3701 non-null   bool   
 6   ReviewBody             3701 non-null   object 
 7   TypeOfTraveller        2930 non-null   object 
 8   SeatType               3699 non-null   object 
 9   Route                  2926 non-null   object 
 10  DateFlown              2923 non-null   object 
 11  SeatComfort            3585 non-null   float64
 12  CabinStaffService      3574 non-null   float64
 13  GroundService          2855 non-null   float64
 14  ValueForMoney          3700 non-null   float64
 15  Reco

In [11]:
df.describe()

Unnamed: 0  OverallRating  SeatComfort  CabinStaffService  \
count  3701.000000    3696.000000  3585.000000        3574.000000   
mean   1850.000000       4.734307     2.875314           3.254057   
std    1068.531001       3.175306     1.362081           1.489709   
min       0.000000       1.000000     1.000000           1.000000   
25%     925.000000       2.000000     2.000000           2.000000   
50%    1850.000000       4.000000     3.000000           3.000000   
75%    2775.000000       8.000000     4.000000           5.000000   
max    3700.000000      10.000000     5.000000           5.000000   

       GroundService  ValueForMoney  Food&Beverages  InflightEntertainment  \
count    2855.000000    3700.000000     3315.000000            2551.000000   
mean        2.784238       2.688378        2.713122               2.646021   
std         1.457058       1.471887        1.437092               1.398509   
min         1.000000       1.000000        1.000000               1.000000   
25%         1.000000       1.000000        1.000000               1.000000   
50%         3.000000       3.000000        3.000000               3.000000   
75%         4.000000       4.000000        4.000000               4.000000   
max         5.000000       5.000000        5.000000               5.000000   

       Wifi&Connectivity  
count         609.000000  
mean            1.924466  
std             1.352705  
min             1.000000  
25%             1.000000  
50%             1.000000  
75%             3.000000  
max             5.000000

In [12]:
#!pip install -q autoviz
#!pip install -q matplotlib pandas numpy scikit-learn

#import textblob
#import textblob.download_corpora
#textblob.download_corpora.download_all()

In [13]:
from autoviz.AutoViz_Class import AutoViz_Class
AV = AutoViz_Class()

report = AV.AutoViz(
    "",
    dfte=df,
    depVar="",
    verbose=2,
    save_plot_dir="figures"
)

Imported v0.1.905. Please call AutoViz in this sequence:
    AV = AutoViz_Class()
    %matplotlib inline
    dfte = AV.AutoViz(filename, sep=',', depVar='', dfte=None, header=0, verbose=1, lowess=False,
               chart_format='svg',max_rows_analyzed=150000,max_cols_analyzed=30, save_plot_dir=None)
Shape of your Data Set loaded: (3701, 20)
#######################################################################################
######################## C L A S S I F Y I N G  V A R I A B L E S  ####################
#######################################################################################
Classifying variables in data set...
  Printing up to 30 columns (max) in each category:
    Numeric Columns : ['OverallRating', 'SeatComfort', 'CabinStaffService', 'GroundService', 'ValueForMoney', 'Food&Beverages', 'InflightEntertainment', 'Wifi&Connectivity']
    Integer-Categorical Columns: []
    String-Categorical Columns: ['TypeOfTraveller', 'SeatType']
    Factor-Categorical Colu

Number of All Scatter Plots = 36


[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to
[nltk_data]    |     /Users/brandonacosta/nltk_data...
[nltk_data]    |   Package cmudict is already up-to-date!
[nltk_data]    | Downloading package gazetteers to
[nltk_data]    |     /Users/brandonacosta/nltk_data...
[nltk_data]    |   Package gazetteers is already up-to-date!
[nltk_data]    | Downloading package genesis to
[nltk_data]    |     /Users/brandonacosta/nltk_data...
[nltk_data]    |   Package genesis is already up-to-date!
[nltk_data]    | Downloading package gutenberg to
[nltk_data]    |     /Users/brandonacosta/nltk_data...
[nltk_data]    |   Package gutenberg is already up-to-date!
[nltk_data]    | Downloading package inaugural to
[nltk_data]    |     /Users/brandonacosta/nltk_data...
[nltk_data]    |   Package inaugural is already up-to-date!
[nltk_data]    | Downloading package movie_reviews to
[nltk_data]    |     /Users/brandonacosta/nltk_data...
[nltk_dat

All Plots are saved in figures/AutoViz
Time to run AutoViz = 25 seconds 


The column where the main text lies is "ReviewBody", which is the column that we will use for data description and further analysis. 

In [15]:
text_len = df['ReviewBody'].fillna("").astype(str).str.len()

min_len = text_len.min()
max_len = text_len.max()
median_len = text_len.median()
mean_len = round(text_len.mean(),2)

data_description = {
    'mininimum text length': min_len,
    'maximum text length': max_len,
    'mean text length':mean_len,
    'median text length':median_len
}
pd.DataFrame([data_description])

mininimum text length  maximum text length  mean text length  \
0           83                   3529               881.28        

   median text length  
0         736.0

In [16]:
examples = df["ReviewBody"].dropna().sample(10, random_state=42)

for i, txt in enumerate(examples, start=1):
    print("\nExample", i)
    print(txt)


Example 1
London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.

Example 2
London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.

Example 3
London to Manchester

In [17]:
plt.figure(figsize=(8,4))
plt.hist(text_len, bins=50)
plt.title("ReviewBody Text Length")
plt.xlabel('characters')
plt.ylabel('frequency')
plt.show()

Step 3: Text Cleaning and Preprocessing

Because VADER uses punctuation/caps as signal, I will preserve a separate cleaning version called text_vader. For the full standard cleaning, a version called text_clean will be used for other models. 

In [19]:
df["text_raw"] = df["ReviewBody"].fillna("").astype(str)

In [20]:
#remove html and numbers
import re

df["text_no_html"] = df["text_raw"].str.replace(r"<.*?>", " ", regex=True)

In [21]:
#remove urls
df["text_no_urls"] = df["text_no_html"].str.replace(r"http\S+|www\.\S+", " ", regex=True)

In [22]:
#remove contractions using library
!pip install -q contractions
import contractions

df["text_expanded"] = df["text_no_urls"].apply(contractions.fix)

In [23]:
#create vader version before addressing capitilization and punctuation
df["text_vader"] = df["text_expanded"].str.replace(r"\s+", " ", regex=True).str.strip()

In [24]:
#create standard cleaned version by recoving whitespace, and making all lowercase
df["text_clean"] = df["text_expanded"].str.lower()

# remove non-letters (keep spaces)
df["text_clean"] = df["text_clean"].str.replace(r"[^a-z\s]", " ", regex=True)

# remove extra whitespace
df["text_clean"] = df["text_clean"].str.replace(r"\s+", " ", regex=True).str.strip()

Now that we have successfully cleaned the data, we will verify the cleaning was successful by running a before/after preview of 3 sample values. 

In [26]:
df[["text_raw", "text_vader", "text_clean"]].sample(3, random_state=42)

text_raw                                                                                                                                                                                                                                                                                                      \
1042                                                                                                                                                                                                                                               London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.   
2460  London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.   
315                                                                                                                                                                                                                                                                                                                                                                                            London to Manchester. British Airways is probably the worst air company I have ever flown with they took 2 hours to unload the luggage and the flight was delayed by an hour. The staff were no help and frankly rude.   

                                                                                                                                                                                                                                                                                                          text_vader                                                                                                                                                                                                                                                                                                       \
1042                                                                                                                                                                                                                                                   London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.   
2460  London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it is the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's did not work in 2 seats - I could not be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I could not even sleep with ear plugs in.   
315                                                                          

Now that we have completed the preliminary preprocessing and necessary data cleaning steps for the dataset, we will proceed with checking the data for any edge cases, like empty strings, extremely short text, and non-english text. 

In [28]:
#empty strings
(df["text_clean"].str.len() == 0).sum()

0

In [29]:
#super short text (anything under 3 words)
word_count = df["text_clean"].str.split().str.len()
(word_count <= 2).sum()

0

In [30]:
#check for non-english text
ascii_ratio = df["text_raw"].apply(lambda x: sum(ord(c) < 128 for c in x) / max(len(x), 1))
(df[ascii_ratio < 0.9].shape[0])

0

In [31]:
#punctuation analysis
has_quotes = df["text_raw"].str.contains(r"[\"“”‘’]", regex=True, na=False)
has_excess_punct = df["text_raw"].str.contains(r"[!?.]{3,}", regex=True, na=False)

print("contains quotes:", has_quotes.sum())
print("has exccessive punctuation:", has_excess_punct.sum())

contains quotes: 798
has exccessive punctuation: 25


In [32]:
df.loc[df["text_raw"].str.contains(r"[!?.]{3,}", regex=True, na=False), "text_raw"].head(5).tolist()

["1. Ground crew in Heathrow left 60 passengers luggage. We had 3 bags. 2. The Heathrow ground crew left 1 of ours behind after the 1st two were received the next day. 3. The Heathrow ground crew lost our 3rd bag 4 times and we were on the phone with them for over 3 hrs over the course of the week! 4. Due to this, we had to spend $500 on health and beauty items for my wife on our cruise. 5. British Airways said they MIGHT reimburse after 3 months. 6. We had to spend $100 in cab fare to Barcelona because the Bacelona ground crew (Iberia) claimed they had delivered it to us. 7. Heathrow ground crew made us 15 minutes late arriving back into Heathrow after our cruise. We actually landed 10 minutes early, but those incompetent people still were not ready for us. 8. The Heathrow ground crew caused us to leave 45 minutes late even though ALL passengers were boarded and ready 15 minutes before the departure time. This will probably mean we will miss our connection in Anchorage. BA clearly doe

Step 4 & 5: Tokenize, Feature Engineer, and Vectorize

Starting with VADER pipeline: 

In [34]:
!pip install vaderSentiment

import vaderSentiment

from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
sia = SentimentIntensityAnalyzer()

In [131]:
def vader_score(text):
    return sia.polarity_scores(str(text))  # returns neg/neu/pos/compound

def vader_label(compound):
    if compound >= 0.05:
        return "positive"
    elif compound <= -0.05:
        return "negative"
    else:
        return "neutral"

import time

start = time.time()

# --- your existing VADER code ---
df['vader_score'] = df['text_vader'].apply(vader_score)

vader_df = df["vader_score"].apply(pd.Series)

df["compound_score"] = vader_df["compound"]
df["positive_score"] = vader_df["pos"]
df["neutral_score"] = vader_df["neu"]
df["negative_score"] = vader_df["neg"]

df["vader_pred"] = df["compound_score"].apply(vader_label)
# --- end VADER code ---

vader_time = time.time() - start
print("VADER total time:", round(vader_time, 3), "sec")
print("Rows processed:", len(df))
print("Time per row:", round(vader_time / len(df), 6), "sec")

VADER total time: 2.126 sec
Rows processed: 3701
Time per row: 0.000574 sec


Vader tokenizes through simple word-level processing with punctuation handling. It uses a broad array of features, like negation, capitlization empphasis, punctuation intensity, and degree modifiers like "very". VADER expects minimal cleaning, since capitalization and punctuations can be preserved. 

TextBlob Pipeline: 

In [37]:
#!pip install -q textblob
from textblob import TextBlob

In [133]:
from textblob import TextBlob

import time

start = time.time()

# --- your existing TextBlob code ---
def textblob_scores(text):
    s = TextBlob(str(text)).sentiment
    return pd.Series([s.polarity, s.subjectivity])

tb_df = df["text_clean"].apply(textblob_scores)
tb_df.columns = ["tb_polarity", "tb_subjectivity"]

df["tb_polarity"] = tb_df["tb_polarity"]
df["tb_subjectivity"] = tb_df["tb_subjectivity"]

def textblob_label(p):
    if p > 0.05:
        return "positive"
    elif p < -0.05:
        return "negative"
    else:
        return "neutral"

df["tb_pred"] = df["tb_polarity"].apply(textblob_label)
# --- end TextBlob code ---

tb_time = time.time() - start
print("TextBlob total time:", round(tb_time, 3), "sec")
print("Rows processed:", len(df))
print("Time per row:", round(tb_time / len(df), 6), "sec")

TextBlob total time: 1.193 sec
Rows processed: 3701
Time per row: 0.000322 sec


TextBlob tokenizes with word-level tokenization. It scores the polarity/subectivity of each word to determine its sentiment. Standard cleaning is expected for this model, with lowercasing and removal of extra punctuations needed. 

Transformer Pipeline:

In [135]:
import torch
import torch.nn.functional as F
import time
from transformers import AutoTokenizer, AutoModelForSequenceClassification

tokenizer = AutoTokenizer.from_pretrained("bhadresh-savani/distilbert-base-uncased-emotion")
model = AutoModelForSequenceClassification.from_pretrained("bhadresh-savani/distilbert-base-uncased-emotion")
model.eval()

labels = ["sadness", "joy", "love", "anger", "fear", "surprise"]

def get_emotion_scores(text):
    inputs = tokenizer(
        str(text),
        return_tensors="pt",
        truncation=True,
        max_length=512
    )
    with torch.no_grad():
        outputs = model(**inputs)
    probs = F.softmax(outputs.logits, dim=1).squeeze().tolist()
    return dict(zip(labels, probs))

start = time.time()

emotion_scores = df["text_clean"].head(200).apply(get_emotion_scores)

emotion_df = emotion_scores.apply(pd.Series)
df.loc[df.index[:200], labels] = emotion_df.values

# add prediction (highest emotion score)
df.loc[df.index[:200], "transformer_pred"] = df.loc[df.index[:200], labels].idxmax(axis=1)

tr_time = time.time() - start
print("Transformer total time:", round(tr_time, 3), "sec")
print("Rows processed:", 200)
print("Time per row:", round(tr_time / 200, 6), "sec")

df[["text_clean"] + labels + ["transformer_pred"]].head(10)

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

Transformer total time: 7.21 sec
Rows processed: 200
Time per row: 0.036052 sec


text_clean                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 hours before takeoff we received a mail stating a cryptic message that there are disruptions to be expected as there is a limit on how many planes can leave at the same time so did the capacity of the heathrow airport really hit british airways by surprise h before departure anyhow we took the one hour delay so what but then we have been forced to check in our hand luggage i travel only with hand luggage to avoid waiting for the ultra slow processing of the checked in luggage overall h later at home than planed with really no reason just due to incompetent people service level far worse then ryanair and triple the price really never again thanks for nothing   
1                                                                                                                                                   i recently had a delay on british airways from bru to lhr that was due to staff shortages they announced that there was a hour holding delay but they would board us immediately in hopes of clearing the gate and leaving early we had to wait the full hours inside the airplane the plane was old dirty had no power at the seats the staff provided a small bag of pretzels and ml of water to the passengers for hour delay and hour flight there were no options to purchase food or drink there were no entertainment options available i am a oneworld emerald elite member but they do not upgrade members based on status first class lounges at heathrow are overcrowded understaffed and poorly equipped the help desk is completely unhelpful when an error arises with delays and cancellations even when having the top status the avios points system has been devalued to near worthlessness and requires fees to book reward that nearly equal the price of the revenue ticket british has lost its way in recent years and has a moved from a world class airline to a budget airline with much worse service and timeliness than ryanair or easyjet   
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            

We previously defined our edge cases as those with excessive punctuations (25 cases) or quotes (798 cases). 

In [42]:
n = 5  # sample of 5 per group of edge cases

# quotes sample
quotes_sample = df.loc[has_quotes, ["text_raw", "text_vader", "text_clean"]].sample(n, random_state=1).copy()

# excess punctuation sample
punct_sample = df.loc[has_excess_punct, ["text_raw", "text_vader", "text_clean"]].sample(n, random_state=1).copy()

In [43]:
# VADER (quotes)
quotes_vader = quotes_sample.copy()
v_scores = quotes_vader["text_vader"].apply(lambda t: sia.polarity_scores(str(t))).apply(pd.Series)
quotes_vader = pd.concat([quotes_vader[["text_raw"]], v_scores], axis=1)

# VADER (excess punctuation)
punct_vader = punct_sample.copy()
v_scores = punct_vader["text_vader"].apply(lambda t: sia.polarity_scores(str(t))).apply(pd.Series)
punct_vader = pd.concat([punct_vader[["text_raw"]], v_scores], axis=1)

In [44]:
quotes_vader

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \
31                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     

In [45]:
punct_vader

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \
1127                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Flew a now ageing B777 flight BA 0053 to Seattle. Pleasant flight with actually quite decent fillet steak for main meal. OK choice of wine / beer with meal and greeted with glass of Prosecco when boarded. Comfortable seat BUT it's the old bug bear - 8 across with "over the legs" access to the aisle. Fortunately we were in the last row so my wife was in the window seat with direct aisle access. And we could look at each other and not at a total stranger. I know BA are saying that they are introducing direct aisle access for everyone in Club Class but this can't come soon enough - and 8 abreast is not exactly a Premium Product in the present circumstances, is it? Absolutely terrible In Flight Entertainment system - although it was On Demand the aircraft was fitted with a very ancient system: blurry, old screens, awful audio (take your own noise reducing headphones) and a pretty dreadful choice of movies.... FA's great welcoming and gave good service. But the main drawback with BA is their insistence that, having paid for a premium service, we then have to pay handsomely again to pre-reserve our seats (up to £105 per seat per journey!).  Not what I expect from a flag carrier airline. We refuse to pay this extortionate sum - our choices have been noted on the loyalty card scheme and fortunately we have got seats together so far. So, although we had quite a pleasant flight to Seattle with BA in Club, if Virgin Atlantic fly the same route we will choose their Upper Class over Clubworld - we can still reserve our seats for free with them.   
1119  This was a night flight New York JFK to London, so it was difficult to assess the service onboard, although what I experienced was certainly not bad. I think this was a very quiet flight, judging 

In [46]:
# TextBlob (quotes)
quotes_tb = quotes_sample.copy()

def tb_scores(text):
    s = TextBlob(str(text)).sentiment
    return pd.Series([s.polarity, s.subjectivity])

tb_out = quotes_tb["text_clean"].apply(tb_scores)
tb_out.columns = ["polarity", "subjectivity"]

quotes_tb = pd.concat([quotes_tb[["text_raw"]], tb_out], axis=1)

# TextBlob (excess punctuation)
punct_tb = punct_sample.copy()

tb_out = punct_tb["text_clean"].apply(tb_scores)
tb_out.columns = ["polarity", "subjectivity"]

punct_tb = pd.concat([punct_tb[["text_raw"]], tb_out], axis=1)

In [47]:
quotes_tb

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \
31                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     

In [48]:
punct_tb

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \
1127                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Flew a now ageing B777 flight BA 0053 to Seattle. Pleasant flight with actually quite decent fillet steak for main meal. OK choice of wine / beer with meal and greeted with glass of Prosecco when boarded. Comfortable seat BUT it's the old bug bear - 8 across with "over the legs" access to the aisle. Fortunately we were in the last row so my wife was in the window seat with direct aisle access. And we could look at each other and not at a total stranger. I know BA are saying that they are introducing direct aisle access for everyone in Club Class but this can't come soon enough - and 8 abreast is not exactly a Premium Product in the present circumstances, is it? Absolutely terrible In Flight Entertainment system - although it was On Demand the aircraft was fitted with a very ancient system: blurry, old screens, awful audio (take your own noise reducing headphones) and a pretty dreadful choice of movies.... FA's great welcoming and gave good service. But the main drawback with BA is their insistence that, having paid for a premium service, we then have to pay handsomely again to pre-reserve our seats (up to £105 per seat per journey!).  Not what I expect from a flag carrier airline. We refuse to pay this extortionate sum - our choices have been noted on the loyalty card scheme and fortunately we have got seats together so far. So, although we had quite a pleasant flight to Seattle with BA in Club, if Virgin Atlantic fly the same route we will choose their Upper Class over Clubworld - we can still reserve our seats for free with them.   
1119  This was a night flight New York JFK to London, so it was difficult to assess the service onboard, although what I experienced was certainly not bad. I think this was a very quiet flight, judging 

The edge cases demonstrate that because TextBlob operates on the clean version of the dataset without punctuations, it didn't affect the performance much. However, the large amount of punctuations is taken into consideration with VADER since it can affect the sentiment analysis. 

Step 6: Analyze and Review Results

We will extract 100 sample rows from the dataset, run the model on each one, and then extract as a csv file where we can manually input the real_y value. 

In [51]:
n = 100
sample_df = df[["text_raw"]].sample(n, random_state=42).copy()
sample_df = sample_df.reset_index(drop=True)

sample_df.head()

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   London to Manchester. British Airways is probably the worst air company I have ever f

In [52]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
import pandas as pd

sia = SentimentIntensityAnalyzer()

def vader_scores(text):
    s = sia.polarity_scores(str(text))  # dict: neg, neu, pos, compound
    return pd.Series([s["neg"], s["neu"], s["pos"], s["compound"]])

# add score columns
v_out = sample_df["text_raw"].apply(vader_scores)
v_out.columns = ["vader_neg", "vader_neu", "vader_pos", "vader_compound"]
sample_df = pd.concat([sample_df, v_out], axis=1)

# label from compound
def vader_label(comp):
    if comp >= 0.05:
        return "positive"
    elif comp <= -0.05:
        return "negative"
    else:
        return "neutral"

sample_df["vader_pred"] = sample_df["vader_compound"].apply(vader_label)

sample_df[["text_raw", "vader_neg", "vader_neu", "vader_pos", "vader_compound", "vader_pred"]].head()

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.   
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   London to Manchester. British Airways is probably the worst air company I ha

In [53]:
from textblob import TextBlob
import pandas as pd

def tb_scores(text):
    s = TextBlob(str(text)).sentiment
    return pd.Series([s.polarity, s.subjectivity])

tb_out = sample_df["text_raw"].apply(tb_scores)
tb_out.columns = ["tb_polarity", "tb_subjectivity"]
sample_df = pd.concat([sample_df, tb_out], axis=1)

def tb_label(p):
    if p > 0.05:
        return "positive"
    elif p < -0.05:
        return "negative"
    else:
        return "neutral"

sample_df["tb_pred"] = sample_df["tb_polarity"].apply(tb_label)

sample_df[["text_raw", "tb_polarity", "tb_subjectivity", "tb_pred"]].head()

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.   
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   London to Manchester. British Airways is probably the worst air company I ha

In [54]:
import torch
import torch.nn.functional as F
import pandas as pd

labels = ["sadness", "joy", "love", "anger", "fear", "surprise"]

def tr_probs(text):
    inputs = tokenizer(str(text), return_tensors="pt", truncation=True, max_length=512)
    with torch.no_grad():
        logits = model(**inputs).logits
    probs = F.softmax(logits, dim=1).squeeze().tolist()
    return pd.Series(probs)

# add 6 probability columns
tr_out = sample_df["text_raw"].apply(tr_probs)
tr_out.columns = labels
sample_df = pd.concat([sample_df, tr_out], axis=1)

# top emotion + sentiment mapping
sample_df["tr_top_emotion"] = sample_df[labels].idxmax(axis=1)

def emotion_to_sentiment(e):
    if e in ["sadness", "anger", "fear"]:
        return "negative"
    elif e in ["joy", "love"]:
        return "positive"
    else:
        return "neutral"  # surprise

sample_df["tr_pred"] = sample_df["tr_top_emotion"].apply(emotion_to_sentiment)

sample_df[["text_raw"] + labels + ["tr_top_emotion", "tr_pred"]].head()

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.   
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   London to Manchester. British Airways is probably the worst air company I ha

In [55]:
sample_df.to_csv("random_sample_review.csv", index=False)

Now that we have extracted a csv file with empty real_y values, we will manually impute the ground truth sentiments to form a table that we can use to calculate the accuracy. 

In [57]:
from sklearn.metrics import confusion_matrix

# load your filled file (rename if needed)
df_filled = pd.read_csv("filled_random_sample_review.csv")

# clean labels
for col in ["real_y", "vader_pred", "tb_pred", "tr_pred"]:
    df_filled[col] = df_filled[col].astype(str).str.strip().str.lower()

# keep only valid labeled rows
valid = df_filled["real_y"].isin(["positive", "neutral", "negative"])
df_filled = df_filled[valid].copy()

labels = ["negative", "neutral", "positive"]

models = ["vader_pred", "tb_pred", "tr_pred"]

for m in models:
    acc = (df_filled[m] == df_filled["real_y"]).mean()
    print("\n", m, "accuracy:", round(acc, 3))

    cm = confusion_matrix(df_filled["real_y"], df_filled[m], labels=labels)
    print(pd.DataFrame(cm, index=labels, columns=labels))


 vader_pred accuracy: 0.73
          negative  neutral  positive
negative     44        1        22   
neutral       0        0         3   
positive      1        0        29   

 tb_pred accuracy: 0.55
          negative  neutral  positive
negative     24       26        17   
neutral       0        1         2   
positive      0        0        30   

 tr_pred accuracy: 0.82
          negative  neutral  positive
negative     52        2        13   
neutral       0        0         3   
positive      0        0        30   


In [58]:
# assumes df_filled already loaded and has: text_raw, real_y, vader_pred, vader_compound

# SUCCESS = matches real_y
vader_success = df_filled[df_filled["vader_pred"] == df_filled["real_y"]]
vader_success[["text_raw", "real_y", "vader_pred", "vader_compound"]].head(3)

text_raw                                                                                                                                                                                                                                                                                                      \
1  London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.   
2                                                                                                                                                                                                                                                                                                                                                                                           London to Manchester. British Airways is probably the worst air company I have ever flown with they took 2 hours to unload the luggage and the flight was delayed by an hour. The staff were no help and frankly rude.   
3                                                                                                                                                                                                                                                                                                    Domestic Edinburgh-Gatwick. Late by 40 mins but that's not unusual. Flight itself lovely clean quiet made up some time. Crew smart and friendly. Only down side was coming off a bridge to a closed gate to then being herded onto a packed bus. Bags came fairly quickly sometimes at LGW it can be a while!   

    real_y  vader_pred  vader_compound  
1  negative  negative      -0.1376      
2  negative  negative      -0.8822      
3  positive  positive       0.9577

In [119]:
# FAILURE = does not match real_y
vader_fail = df_filled[df_filled["vader_pred"] != df_filled["real_y"]]
vader_fail[["text_raw", "real_y", "vader_pred", "vader_compound"]].sample(3, random_state=1)

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \
80                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                After my flight to Vienna with BA I was hoping for a slight improvement on the way back - however the service was very boring. Drink service was rushed with the stewardess pouring an entire miniature of vodka into a drink before putting a drop of Tomato juice in when I asked for a Tomato and Vodka - unpalatable. Was glad to be off the plane in all fairness. Hoping that one day BA will wow me.   
68  London Heathrow to Krakow. I am a one world emerald member on AA so I get certain perks when flying British Airways such as free seat selection, lounge use and priority boarding - that makes this airline bearable - just. They are obsessed with carry on bags much like Ryanair giving out silly yellow tags for carry on bags and checking carry on sizes - not sure what it does for them as it only increases their bag handling costs. Their seats are no better than anyone else and their pitch just as tight. What used to set BA apart was their inflight service - even in coach a sandwich or something - this has degraded to nothing, zero - even a soft drink is charged. So I asked for water, I was offered an Evian for $5 - no thanks just plain water. Ok sir we will come back to you after we serve everyone else. Of course he never came back - 2.5 hour flight and not even allowed a cup of water. BA are listed as a 4 star airline,  AA are a 3 star yet AA now have better inflight service than BA by a country mile. BA are a bottom tier airline and nothing more to the average flier.   
83                                                                                                                                                                                                                                                                                                                         Food on this 4+ hours flight from Moscow to Heathrow was a filled roll and 1 cup of tea/coffee. The other London-Moscow flights provide a full meal. The cabin staff were almost robotic and programmed in the discharge of their duties with no interaction with passengers. When the flight touched down, it was 75 minutes before we disembarked, 6 British Airways flights that landed after us discharged their passengers whilst our flight waited first at a stand and then 15 minutes for an air bridge operator. My email to their CEO produced nothing except a plastic apology, even though the captain continually said we were waiting on find the guy responsible for stand allocation. Used to be BA was a world leader. They've followed the empire and dinosaurs into extinction.   

     real_y  vader_pred  vader_compound  
80  negative  positive       0.8863      
68  negative  positive       0.5704      
83  negative  positive       0.0772

In [60]:
# SUCCESS = matches real_y
tb_success = df_filled[df_filled["tb_pred"] == df_filled["real_y"]]

tb_success[["text_raw", "real_y", "tb_pred", "tb_polarity", "tb_subjectivity"]].head(3)

text_raw                                                                                                                                                                                                                                                                                                      \
0                                                                                                                                                                                                                                               London Heathrow to Stuttgart. Absolutely disgusted with BA. my husband was asked to put his one piece of hand luggage in the hold due to the flight being busy and he agreed. BA had one job to do - put it in the hold and make sure it arrives in Stuttgart. This has proven too difficult. The bag is lost and they have not offered any assistance or support.   
1  London to New York JFK with British Airways, and abyssal customer service. Onboard the flight can be great, and it's the website and shoddy call centre staff who let them down. However, this time the onboard service was as bad as the ground service. Food was poor. I was in the upper deck with only 5 other business class passengers, and yet my food choice was gone. The TV's didn't work in 2 seats - I couldn't be bothered to try a third and decided to just get some sleep. The staff insisted on forming a cliquish huddle and were speaking so loudly I couldn't even sleep with ear plugs in.   
2                                                                                                                                                                                                                                                                                                                                                                                           London to Manchester. British Airways is probably the worst air company I have ever flown with they took 2 hours to unload the luggage and the flight was delayed by an hour. The staff were no help and frankly rude.   

    real_y   tb_pred   tb_polarity  tb_subjectivity  
0  negative  negative  -0.205000       0.712778      
1  negative  negative  -0.053683       0.452925      
2  negative  negative  -0.433333       0.533333

In [121]:
# FAILURE = does not match real_y
tb_fail = df_filled[df_filled["tb_pred"] != df_filled["real_y"]]

tb_fail[["text_raw", "real_y", "tb_pred", "tb_polarity", "tb_subjectivity"]].sample(3, random_state=1)

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \
14                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          San Diego to Hannover via London. I booked on ba.com, the longer flights both ways in premium economy.  On the return I was informed at San Diego that the American Airlines flight into DFW was severely delayed and I would miss my flight into London. American Airlines could book me on a later flight the same day, but had only  coach available. If I insisted on premier economy I could fly in 3 days. I took the coach and American assured my that BA would refund the fare difference for this downgrade. After arrival at home I contacted BA on their website to refund the fare difference. A bizarre exchange followed. First they said you did not buy any upgraded seat. I told them to look at the ticket. You can't get a refund because the economy class seat was much more expensive that the booked premium economy seat. I contacted BA Germany who said you get a refund, but you booked this with a travel agent, so we are not in charge.  I booked with BA on their website. They saidthey would send this to London so they can compute the refund, but then you can't get a refund because your seat was much more expensive than the premium seat you booked months in advance. The change in booking was the sole fault of the delayed American Airlines flight. This is miserable customer service.   
12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

In [115]:
# SUCCESS = matches real_y
tr_success = df_filled[df_filled["tr_pred"] == df_filled["real_y"]]

tr_success[["text_raw", "real_y", "tr_pred", "tr_top_emotion"]].sample(3, random_state=1)

text_raw                                                                                                                                                                                                                                                                                                                                                                                                              \
11  Flew YVR to MUC via LHR. I had high expectations for this airline but was pretty disappointed. We were flying with twin 15 month olds so it's great with BA that you can preselect the bassinet spots when booking. Crew did bare minimum seemed as though asking for anything was a huge inconvenience. Audio on IFE was awful couldn't hear any movies. Flight landed 10 minutes late in London and we were denied boarding on the MUC flight due to being late. If they allow a connection of 1h10 and you're 10 minutes late and can't get on their system is flawed. When are any LHR flights not 10 mins late! They rebooked us 2.5 hours later but after complaining put us on a Lufthansa flight 1 hour later. I never thought I would say this but will use Air Canada on this route from now on if the price is equal.   
85                                                                                                                                                                                                                                                                                                                                     London to Seville. Following on from mistakenly booking a return flight with BA despite having a previous confirmed booking with a competitor, I quickly called up BA to resolve the issue. They understood the problem and gave me a full refund. The agent was exceptionally helpful on the phone and provided me a quick answer and solution to the problem. I don’t believe most airlines would have accepted my reasoning, however it’s great to see BA treat their customers with care.   
57                                                                                                                 LHR-PEK-LHR. We travel BA club world often and my overall impression is very good. Yes sometimes the attitude of the staff can be not up to expectations or the food not comparable with some of the Asian airlines but the price you pay for a business class seat with BA is unbeatable. And for me a flat bed on an overnight flight is what matters the most. Contrary to general recommendations when I travel with a companion I prefer the middle seats on the main deck and not the upper deck because by using the screen you can have your own little intimate cabin. And if the flight is not full is so much better than the claustrophobic upper deck. Strongly recommended for a discounted ticket.   

     real_y   tr_pred  tr_top_emotion  
11  negative  negative     sadness     
85  positive  positive         joy     
57  positive  positive         joy

In [117]:
# FAILURE = does not match real_y
tr_fail = df_filled[df_filled["tr_pred"] != df_filled["real_y"]]

tr_fail[["text_raw", "real_y", "tr_pred", "tr_top_emotion"]].sample(3, random_state=1)

text_raw                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \
28                             BA362 6th June 2015. If this were the first flight I had taken with British Airways I would not be impressed. At check in there is a lack of professionalism. The customer does not come first even at the First Class check in. Staff interrupt each other and discuss their own business in priority to completing the check in. The flight was in fact delayed by an hour despite the information such as was provided given as "ten minutes". The aircraft appeared dated and not especially clean with crumbs on the seat and grease marks on the windows. The inflight shopping was made difficult as there were no inflight duty free magazines available in Club Class. Inflight entertainment system did not work. Whilst the food was good it had to be changed as the first offering was cold. Inflight crew were pleasant and efficient. Upon arrival the baggage reclaim took no account of priority passengers.   
23                                                                                                                                                                                                                                                                                                                                                                                                                  Leeds Bradford to Heathrow. Service is equal to Ryanair and EasyJet. I've been flying with BA regularly for last 4 years and I can't get my head around what is going on. Complimentary drinks replaced with a Buy on Board menu and the flight is so short not all passengers even get served now!  Really just appalling. There is no competition on the LBA route, but I certainly won't pay anything extra to travel with BA from now on if there are alternatives available - Ryanair it is at around a fifth of the price usually!   
70  Our flight from London to Boston was okay. Disappointed to only be able to sit in middle of row of 10. Annoying to have to pay extra to have a window seat. Also catering on board is disappointing with only one meal plus a biscuit but crew were friendly and helpful. The return flight was different. Staff not particularly helpful. Person in front reclined seat fully and I wasn't able to fit food tray in. Crew member said I was causing problems for the man in front when I requested he put his seat up to eat. Also armrests did not lift up leaving less than 6 inches between seat in front and armrest meaning I was literally trapped in my seat for 6 hours. Not comfortable and unsafe in an emergency. We had one drinks trolley followed by a meal around an hour later. Staff then disappeared and only water or juice available for rest of flight. Will not use BA again. Delta and Virgin flights offer much better service.   

     real_y   tr_pred  tr_top_emotion  
28  negative   neutral    surprise     
23  negative  positive         joy     
70  negative  positive         joy

NameError: name 'vader_pred_text' is not defined